zebra: add IPv6 router-id

sylane opened the pull request:
* add command `[no] ipv6 router-id X:X::X:X [vrf NAME]`.
 * add command `show ipv6 router-id [vrf NAME]`.
 * add command `[no] ip router-id A.B.C.D [vrf NAME]` and
   make the old one without `ip` an alias for it.
 * add command `show ip router-id [vrf NAME]` and make
   the old one without `ip` an alias for it.
 * add ZAPI commands `ZEBRA_ROUTER_ID_V6_ADD`,
   `ZEBRA_ROUTER_ID_V6_DELETE` and `ZEBRA_ROUTER_ID_V6_UPDATE`
   for deamons to get notified of the IPv6 router-id.
 * update zebra documentation.

donaldsharp commented:
I think it would be far preferable to just add a afi to the existing router-id ZAPI messages

sylane commented:
@donaldsharp Are you suggesting to reuse the same ZAPI message to send both prefixes ? Would that still use two callbacks and read functions or are you suggesting to change that too and update all the daemon using the router id callback ?

donaldsharp commented:
I'm suggesting that adding 3 new zapi messges when all you have to add to the original is a AFI is all you need to do

sylane commented:
@donaldsharp The prefix already contains an AFI, we could send both router-id through the same ZAPI message and the client could then call one or the other callback in function of the prefix family. So we wouldn't have any new zapi message, it would only be called multiple times and it would be the client that would select the callback for backward compatibility. Does that sound good ?

sylane commented:
@donaldsharp So the problem with your suggestion is that AFAIK the message dispatcher cannot inspect the payload before calling the daemon callback, so it has no way of knowing if it hsould use the old router id callback or the new IPv6 callback. It means that I would break the API and I would have to change all the router id callback in all the FRR daemon so they ignore the IPv6 prefixes received by the router id update callback.

donaldsharp commented:
To further expand on my point I see the notification events for routerid loosely analagous to the notifcation events for redistribution in which we do not have v4 or v6 specific calls.  I beleive that Renato has outlined how I would actually approach the issue just above as well.

sylane commented:
@donaldsharp @rwestphal I updated the code following your suggestions, I hope it looks better now.

pguibert6WIND commented:
To illustrate, do you have any references to give where a protocol uses ipv6 router id, please?

sylane commented:
@pguibert6WIND I am using this in the pathd daemon, in this branch https://github.com/opensourcerouting/frr/tree/path-pcep/router-id-v6

sylane commented:
@pguibert6WIND It is used to decide on the address a PCC connect from and advertise to a PCE when using IPv6

sylane commented:
@rwestphal I think I fixed all your concerns in my last push

sylane commented:
CI:rerun

sylane commented:
@rwestphal I think I finally got everything right :)

sylane commented:
CI:rerun

rwestphal commented:
CI:rerun

sylane commented:
test_ibgp_ecmp_topo2 pass perfectly locally, I just rebased on top of the last master.

sylane commented:
CI:rerun

donaldsharp commented on lib/zclient.c:
```diff
@@ -554,6 +554,28 @@ void zclient_send_dereg_requests(struct zclient *zclient, vrf_id_t vrf_id)
 	}
 }
 
+int zclient_send_router_id_add(struct zclient *zclient, afi_t afi,
```
let's create a generic function that both the _add and _delete function call.  cut-n-paste code is not good

donaldsharp commented on zebra/router-id.c:
```diff
@@ -59,60 +59,107 @@ static struct connected *router_id_find_node(struct list *l,
+	switch (afi) {
+	case AFI_IP:
+		if (p->u.prefix4.s_addr == 0) /* unset */
+		{
```
{ belongs on the previous line

sylane replied:
This was formatted like this before my changes, I just kept the original formatting and comment. It seems clang is formatting the code like this due to the comment, I'll remove it.

donaldsharp commented on zebra/router-id.c:
```diff
@@ -59,60 +59,107 @@ static struct connected *router_id_find_node(struct list *l,
+		if (p->u.prefix4.s_addr == 0) /* unset */
+		{
+			zvrf = vrf_info_lookup(vrf_id);
+			if (!zvrf)
```
let's move the zvrf lookup and check for happiness outside of the switch statement it's the same for v4 or v6

sylane replied:
I don't think it can be moved out as this is looked up only if the address is not set, and how to check it depends on the address type.

donaldsharp commented on zebra/router-id.c:
```diff
@@ -59,60 +59,107 @@ static struct connected *router_id_find_node(struct list *l,
+		break;
+	case AFI_IP6:
+		if (router_id_v6_is_any(p)) /* unset */
+		{
```
{ belongs on the previous line

donaldsharp commented on zebra/router-id.c:
```diff
@@ -155,134 +217,264 @@ void router_id_del_address(struct connected *ifc)
+				vty_out(vty, "ip router-id %pI4\n",
+					&zvrf->rid_user_assigned.u.prefix4);
+			else
+				vty_out(vty, "ip router-id %pI4 vrf %s\n",
```
The vrf specific router id's belong under the vrf section.

ip router-id .....  -> default vrf

vrf BLUE
  ip router-id ....
!
vrf GREEN
  ip router-id ....
!

sylane replied:
What about backward compatibility ?

donaldsharp commented on zebra/router-id.c:
```diff
@@ -155,134 +217,264 @@ void router_id_del_address(struct connected *ifc)
-DEFUN (router_id,
+DEFUN (ip_router_id,
+       ip_router_id_cmd,
+       "ip router-id A.B.C.D [vrf NAME]",
```
vrf NAME should be implied by the vrf subsection we are in.

sylane replied:
Are you suggesting to remove the optional argument `vrf` from the new commands `ip router-id` and `ipv6 router-id`, and keep it in the old `router-id` one ? Or to break the backward compatibility and remove it from all the router-id commands ?

donaldsharp commented on zebra/router-id.c:
```diff
@@ -155,134 +217,264 @@ void router_id_del_address(struct connected *ifc)
+       "Show the configured router-id\n"
+       VRF_CMD_HELP_STR)
+{
+	int idx = 0;
```
This is cut-n-paste code, it cannot be abstracted?

sylane commented:
@donaldsharp Are you suggesting to remove the optional argument vrf from the new commands ip router-id and ipv6 router-id, and keep it in the old router-id one ? Or to break the backward compatibility and remove it from all the router-id commands ?

sylane commented:
@donaldsharp I think I addressed most of your concerns minus changing the way the VRF parameter is handled. I think this is out of the scope of this PR. Even though I agree this is a change that make sense, it is not what this PR is about, the PR is about adding an IPv6 router identifier, not changing how the router id commands handle the VRF.

donaldsharp commented:
Please rebase ontop of https://github.com/FRRouting/frr/pull/6706

I'm more than a bit annoyed about the time spent arguing about putting into a submode when it took me < 30 minutes to program it myself.
